### PR TITLE
refactor: support bitmap for u8/16 and i8/16 in `approx_distinct`

### LINKED ISSUE #1109: `approx_distinct` should be leveraging bitmap for counting u8/16 and i8/16
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

`approx_distinct` should be leveraging bitmap for counting u8/16 and i8/16

**Describe the solution you'd like**

Using a bitmap is more efficient 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Follow #1841
Closes #1109

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/approx_distinct.rs [hunk 1]
```diff
@@ -304,5 +317,89 @@ where
...
+        }
+        Ok(())
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        Ok(ScalarValue::from(self.bitmap.len()))
+    }
+
+    fn size(&self) -> usize {
+        self.bitmap.serialized_size()
```
Comment (Weijun-H) [inline datafusion/physical-expr/src/aggregate/approx_distinct.rs:403]: Not sure is a proper way to measure roaring bitmap

**Comment by korowa** [comment]
Thank you @Weijun-H!

This PR, though, looks like the implementation of `CountDistinct` accumulator, and it doesn't seem that any approximation is performed in the code -- won't it be better to use these changes as default `CountDistinct` implementation for specified data types?

**Comment by alamb** [comment]
I wonder what we should do with this PR now we have
* https://github.com/apache/arrow-datafusion/pull/8849
* https://github.com/apache/arrow-datafusion/pull/8721

**Comment by korowa** [comment]
> I wonder what we should do with this PR now we have

While I'm still not sure if this PR fits `ApproxDistinct` functionality, I suppose it might be a viable replacement for HashSets in regular `CountDistinct` -- so it at least worth checking / benchmarking within #1823 (paying special attention to memory consumption of bitmap-based accumulator)

**Comment by alamb** [comment]
Thanks @Weijun-H  and @korowa  -- I'll mark this PR as a draft now and if someone finds time to do the benchmarks we can reopen it with that in consideration

**Pull request closed without merging**